Fix deadlock in AsyncOperation when TryCancel() called from TryComplete()

shaan1337 opened the pull request:
Fixes https://github.com/dotnet/corefx/issues/42718

This is an attempt to fix above deadlock issue. I've opened a draft PR to obtain feedback on whether the approach looks correct. Thoughts are welcome.

Since `AsyncOperation` has private visibility, I've not added any failing tests. I'm now attempting to test the patched assembly on Azure Devops to see if our tests still hang on the macOS hosted agent.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=381) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=381) before we can accept your contribution.<br/><br/>:x: shaan1337 [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=381)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=381) it.</sub>

tmds commented:
> I've opened a draft PR to obtain feedback on whether the approach looks correct. Thoughts are welcome.

I think we should add an additional `State.Aborted` that doesn't do the spinning which is needed `State.Cancelled`.
For cancelling we need to spin, because we don't want cancel a Running operation that was successful.
This spinning isn't needed when the Socket is closed. After the running operation is done, we can CompareExchange to see if the operation got aborted.

cc @scalablecory @stephentoub @geoffkizer

shaan1337 committed:
Fix deadlock in AsyncOperation when op.TryCancel() is called from op.TryComplete():
- A socket closure can trigger socketAsyncContext.StopAndAbort() from the same thread where the operation is running, inside TryComplete()
- StopAndAbort() subsequently calls TryCancel()
- The operation ends up waiting for itself since it keeps spinning until the operation state changes from 'Running' to 'Waiting' (which never happens)

Fix description: Immediately abort operations (if any) when the socket is closed
- Add a boolean flag `abort` to TryCancel() to specify that we want to immediately abort operations and set it to `true` when calling TryCancel() from StopAndAbort()
- Add a new state: State.Aborted
- Directly transition from `Running` or `Waiting` to `Aborted` if `abort` is set instead of spinning
- Modify SetComplete(), SetWaiting() and TrySetRunning() to take new `Aborted` state into consideration
- Simplify transition code & TryCancel() code

Co-Authored-By: Tom Deseyn <tom.deseyn@gmail.com>

shaan1337 commented:
@tmds thanks for the pointers, I've tried to rewrite the fix based on your solution.

halter73 commented:
How difficult would it be to write a regression test for this?

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -178,21 +179,35 @@ public bool TrySetRunning()
             {
-                Debug.Assert(Volatile.Read(ref _state) == (int)State.Running);
+                State oldState = (State)Interlocked.CompareExchange(ref _state, (int)State.Complete, (int)State.Running);
+                if (oldState == State.Aborted)
```
suggestion: `return oldState == State.Running`.

tmds replied:
and maybe before that `Debug.Assert(oldState == State.Running || oldState == State.Aborted)`.

tmds replied:
`TrySetRunning` needs to be updated also.

shaan1337 replied:
@tmds  this looks better, i'll make the changes thanks.

For `TrySetRunning`, it seems that changes are not required there if we always only transition from `Running` to `Aborted`. However, if we can also transition from `Waiting` to `Aborted`, then it would need to be changed.

When writing the fix, I tried to define `Aborted` and `Cancelled` as follows which sounds consistent. It might need some discussion:
`Aborted` - an operation is (ungracefully) stopped during its execution
`Cancelled` - an operation is stopped before its even executed or it's idle waiting (which means it's been gracefully stopped)

what do you think?

shaan1337 replied:
@tmds i've made the changes, also allowing transitions from any state to `Aborted`

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -210,11 +225,20 @@ public bool TryCancel()
+                                // Otherwise, the next iteration of the loop will handle the new state.
+                                Trace("Operation aborted");
+                                keepWaiting = false;
+                            }
```
I tried splitting this switch statement up in the two things it is doing: change state, and exit the function if the oldstate was already completed. wdyt?

```cs
State oldState;
if (aborting)
{
    oldState = (State)Interlocked.Exchange(ref _state, (int)State.Aborted);
}
else
{
    // Cancel operation when it is not completed.
    SpinWait spinWait = default;
    do
    {
        oldState = (State)Interlocked.CompareExchange(ref _state, (int)State.Cancelled, (int)State.Waiting);

        if (oldState == State.Running)
        {
            // A completion attempt is in progress. Keep busy-waiting.
            Trace("Busy wait");
            spinWait.SpinOnce();
        }
    } while (oldState == State.Running);
}

// Exit if the operation had already completed.
switch (oldState)
{
    case State.Complete:
        Trace("Exit, previously completed");
        return false;
    case State.Cancelled:
    case State.Aborted:
        Trace("Exit, previously cancelled/aborted");
        return false;
}
```

shaan1337 replied:
@tmds this looks better thanks, i'll make the changes asap.

there's just one thing i'm not sure about, which still revolves around the definition of `Aborted` vs `Cancelled` as mentioned [here](https://github.com/dotnet/runtime/pull/381/files/dd0ccf0469ae53707a0fb4e456dac0abc8e3b08f#r354797209):

When [aborting](https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L1111) there'll be at most one operation on the queue that's actually in the `Running` state. The others will all be in `Waiting` state. So, with the current code, only the running operation will be aborted and the others will be cancelled which is consistent with the above [definition](https://github.com/dotnet/runtime/pull/381/files/dd0ccf0469ae53707a0fb4e456dac0abc8e3b08f#r354797209) : The running operation was aborted since it was "ungracefully" stopped while the waiting operations on the queue would be cancelled since they've not started to execute.

shaan1337 commented:
@halter73 it seems that this test is very similar to the scenario: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs#L900

I tried to reproduce the issue on linux with a few modifications to this test but couldn't. We saw the problem occurring primarly on macOS 10.13 hosted agent on Azure pipelines but unfortunately I do not have a proper mac environment set up to try to reproduce it.

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -793,7 +788,7 @@ public bool StartAsyncOperation(SocketAsyncContext context, TOperation operation
                                 if (cancellationToken.CanBeCanceled)
                                 {
-                                    operation.CancellationRegistration = cancellationToken.UnsafeRegister(s => ((TOperation)s).TryCancel(), operation);
+                                    operation.CancellationRegistration = cancellationToken.UnsafeRegister(s => ((TOperation)s).TryCancel(false), operation);
```
nit: `TryCancel(aborting: false)`

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -1106,7 +1106,7 @@ public bool StopAndAbort(SocketAsyncContext context)
                         do
                         {
-                            aborted |= op.TryCancel();
+                            aborted |= op.TryCancel(true);
```
nit: `TryCancel(aborting: true)`

tmds commented:
I think the fix is good.

I'm trying to understand how you get in this situation. I wonder if you are Disposing the SocketSafeHandle?

Usually, you should Dispose the Socket. Disposing the Socket leads to SocketAsyncContext.Close (InnerReleaseHandle) on the Socket.Dispose thread:
```cs
        internal void CloseAsIs()
        {
            if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"_innerSocket={_innerSocket}");

#if DEBUG
                // If this throws it could be very bad.
            try
            {
#endif
                InnerSafeCloseSocket innerSocket = _innerSocket == null ? null : Interlocked.Exchange<InnerSafeCloseSocket>(ref _innerSocket, null);

                Dispose();
                if (innerSocket != null)
                {
                    // Wait until it's safe.
                    SpinWait sw = new SpinWait();
                    while (!_released)
                    {
                        sw.SpinOnce();
                    }

                    InnerReleaseHandle();
```
In your stacktrace, the SocketAsyncContext Close is due to the SafeHandle release:
```cs
        protected override bool ReleaseHandle()
        {
            if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"_innerSocket={_innerSocket}");

            _released = true;
            InnerSafeCloseSocket innerSocket = _innerSocket == null ? null : Interlocked.Exchange<InnerSafeCloseSocket>(ref _innerSocket, null);
            if (innerSocket != null)
            {
#if DEBUG
                // On AppDomain unload we may still have pending Overlapped operations.
                // ThreadPoolBoundHandle should handle this scenario by canceling them.
                innerSocket.LogRemainingOperations();
#endif

                InnerReleaseHandle();
                innerSocket.DangerousRelease();
            }
```
If Socket.Dispose was called (CloseAsIs), _innerSocket should be null.

shaan1337 commented:
@tmds I've checked through the code but it does not appear we're dealing directly with the SafeSocketHandle. I'll try to check again asap.

tmds commented:
@shaan1337 I'm trying to understand why your program is not taking the `Socket.Dispose` calls `SocketAsyncContext.Close` path. Do you know what code is using this `Socket`? Do you call `Socket.Dispose` for this `Socket`?

The PR itself is good for me.

thefringeninja commented:
@tmds I did much of the work moving our solution to dotnet core. In fact I had added `.Dispose()` there and it did not seem to make a difference.

shaan1337 commented:
@tmds I'm not sure how SafeHandles works under the hood or if the following makes sense: What would happen if the socket file descriptor is closed internally from the `recvfrom` syscall by the kernel before it returns to managed code? Would the SafeHandle's `ReleaseHandle` function be called immediately on the same thread when returning to managed code?

halter73 commented:
I'm not sure how the SafeHandle would know the underlying file descriptor was closed unless there's some Socket-specific logic to do so. Generally speaking, SafeHandle.ReleaseHandle() is only called if SafeHandle.Dispose(), SafeHandle.DangerousRelease() and/or the SafeHandle finalizer is invoked.

tmds commented:
> In fact I had added .Dispose() there and it did not seem to make a difference.

@thefringeninja are you Disposing the Socket (**not** the SafeHandle)? What .NET Core version are you using that is giving the issue?

shaan1337 commented:
@tmds apologies for the delay, I can confirm that the issue also occurs on .NET core 3.1.100 (still on macOS 10.13)

Our code usually follows this pattern:
```
socket.Shutdown(SocketShutdown.Both)
socket.Close(timeout)
```

I've tried to change it to the following everywhere in the code base:
```
socket.Shutdown(SocketShutdown.Both)
socket.Close(timeout)
socket.Dispose()
```

However, the issue still occurs.

I'm trying to get a minimal reproduction of the issue but it does not seem straightforward. If there's any tracing information that could be helpful to obtain the code path please let me know.

thefringeninja commented:
Disposing the Socket.

On Tue, Dec 17, 2019 at 2:45 AM Tom Deseyn <notifications@github.com> wrote:

> In fact I had added .Dispose() there and it did not seem to make a
> difference.
>
> @thefringeninja <https://github.com/thefringeninja> are you Disposing the
> Socket (*not* the SafeHandle)? What .NET Core version are you using that
> is giving the issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/runtime/pull/381?email_source=notifications&email_token=AADY7B5A7YXZ64BOXGYTE4TQZCUTJA5CNFSM4JST6GT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHB6UCQ#issuecomment-566487562>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADY7B2PFD5LTXWGMTF4P6TQZCUTJANCNFSM4JST6GTQ>
> .
>


-- 
Sent from my regular computer
http://twitter.com/thefringeninja
http://www.thefringeninja.com/

shaan1337 commented:
Minimal repro: https://github.com/dotnet/corefx/issues/42718#issuecomment-569046141

I tried different combinations here: https://github.com/shaan1337/system-net-sockets-deadlock-bug-42718/blob/master/system_net_sockets_deadlock_bug_42718.cs#L28 but only this one worked so far. It appears that `Shutdown(SocketShutdown.Both)` followed by `Close()` with a timeout causes the issue (it didn't hang when calling Close without a timeout)

tmds commented:
> It appears that Shutdown(SocketShutdown.Both) followed by Close() with a timeout causes the issue (it didn't hang when calling Close without a timeout)

And only on mac, right?

Can you add a test for this?

shaan1337 commented:
> And only on mac, right?

yes, we've tested on macOS 10.13 and 10.14, they are both affected

> Can you add a test for this?

ok, i'll do it asap

shaan1337 committed:
Add regression test for dotnet/corefx#42718 issue

shaan1337 commented:
@tmds I've added the regression test. Unfortunately, it seems that the fix is not working since the regression test seems to fail on macOS Debug x64:  https://dev.azure.com/dnceng/public/_build/results?buildId=481177

It could be an issue with the test as well.

I've downloaded the logs from helix and I can see the following:
```   System.Net.Sockets.Tests.SendReceiveApm.ReceiveAsync_ConcurrentShutdownWithTimeoutAndClose_SucceedsOrThrowsAppropriateException [FAIL]
      System.Net.Internals.SocketExceptionFactory+ExtendedSocketException : Operation timed out 127.0.0.1:61534
      Stack Trace:
        /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(4279,0): at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
        /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(787,0): at System.Net.Sockets.Socket.Connect(EndPoint remoteEP)
        /_/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(984,0): at System.Net.Sockets.Tests.SendReceive`1.CreateConnectedSocketPair()
        /_/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(945,0): at System.Net.Sockets.Tests.SendReceive`1.ReceiveAsync_ConcurrentShutdownWithTimeoutAndClose_SucceedsOrThrowsAppropriateException()
        --- End of stack trace from previous location ---
```
Link: https://helix.dot.net/api/2019-06-17/jobs/522890ec-cbce-44ab-96bb-6d5cf5a3704b/workitems/System.Net.Sockets.Tests/files/console.0b84be8d.log


I've also opened a 2nd PR here, where I've reverted the fix: https://github.com/dotnet/runtime/pull/1669

On this PR, the macOS build did not fail: https://dev.azure.com/dnceng/public/_build/results?buildId=481183
It may have just been luck, we may need to run it a few times to see if it fails

I do not have the authorization to relaunch the builds, can someone please do it? thanks

fehmianac commented:
Hi All,

We also same issue in Production stage. 

Situation like that.
We are doing "fail over" our MSSQL database. Some time we had dead lock thread after fail over. When we analyze dump we are seeing some dead lock in "system.net.sockets.socketpal.receive" method. 

Is there any work around about this case ? 


OS: Ubuntu 18.04 also Docker
Runtime: 3.1.100
ORM: Dapper 
Data Client: System.Data.SqlClient 4.80
RPM: About 10K

davidsh commented:
@shaan1337 This PR now has a conflict. Can you please rebase with PR with latest 'master'? Doing so will also start another round of CI testing.

davidsh commented:
Thank you for the PR.

We're closing the PR at this time due to inactivity.

When you're ready to resume work on this, the PR can be re-opened.
